File: backend/src/Gorodki.Domain/Geo/Geodesy.cs
namespace Gorodki.Domain.Geo;

/// <summary>
/// Расстояния на поверхности Земли — та же формула и тот же порядок действий, что в <c>Geodesy</c> из GameCore (Swift):
/// судья отрезков на сервере должен получить те же числа, что телефон.
/// </summary>
public static class Geodesy
{
    /// <summary>Средний радиус Земли по IUGG, метры.</summary>
    public const double MeanEarthRadius = 6_371_008.8;

    /// <summary>Расстояние по дуге большого круга (формула гаверсинусов), метры.</summary>
    public static double Distance(double latitude1, double longitude1, double latitude2, double longitude2)
    {
        var lat1 = Radians(latitude1);
        var lat2 = Radians(latitude2);
        var deltaLat = Radians(latitude2 - latitude1);
        var deltaLon = Radians(longitude2 - longitude1);

        var h =
            (Math.Sin(deltaLat / 2) * Math.Sin(deltaLat / 2))
            + (Math.Cos(lat1) * Math.Cos(lat2) * Math.Sin(deltaLon / 2) * Math.Sin(deltaLon / 2));
        return 2 * MeanEarthRadius * Math.Asin(Math.Min(1, Math.Sqrt(h)));
    }

    private static double Radians(double degrees) => degrees * Math.PI / 180;
}

File: ios/Packages/GameCore/Sources/GameCore/Geo/Geodesy.swift
import Foundation

/// Расстояния на поверхности Земли.
public enum Geodesy {
    /// Средний радиус Земли по IUGG, метры.
    public static let meanEarthRadius = 6_371_008.8

    /// Расстояние по дуге большого круга (формула гаверсинусов), метры.
    ///
    /// Точность — доли процента: Земля не шар. Для отрезков GPS-следа (метры и десятки метров)
    /// этого более чем достаточно. Точную площадь участков считает сервер в проекции UTM.
    public static func distance(from a: Coordinate, to b: Coordinate) -> Double {
        let lat1 = a.latitude.radians
        let lat2 = b.latitude.radians
        let deltaLat = (b.latitude - a.latitude).radians
        let deltaLon = (b.longitude - a.longitude).radians

        let h =
            sin(deltaLat / 2) * sin(deltaLat / 2)
            + cos(lat1) * cos(lat2) * sin(deltaLon / 2) * sin(deltaLon / 2)
        return 2 * meanEarthRadius * asin(min(1, h.squareRoot()))
    }
}

extension Double {
    /// Градусы → радианы.
    var radians: Double { self * .pi / 180 }
    /// Радианы → градусы.
    var degrees: Double { self * 180 / .pi }
}
